scripts/runSonarQube.sh

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+
+# Check for install/updates at https://github.com/SonarSource/sonarqube
+
+sonar_host="http://localhost:9000"
+sonar_project="benchmark"
+sonar_token="291e50b66b5f5198e6ac0d49bf1057ac3a3a0fc8"
+
+mvn sonar:sonar \
+  -Dsonar.projectKey="$sonar_project" \
+  -Dsonar.host.url="$sonar_host" \
+  -Dsonar.login="$sonar_token"
+
+sleep 300s # might be replaced with polling of $sonar_host/api/ce/component?component=$sonar_project
+
+benchmark_version=$(scripts/getBenchmarkVersion.sh)
+sonarqube_version=$(curl --silent -u "$sonar_token:" "$sonar_host/api/server/version")
+result_file="results/Benchmark_$benchmark_version-sonarqube-v$sonarqube_version.json"
+
+# SonarQube does not provide a download option so we've to create the result file manually :(
+
+result='{"issues":[], "hotspots": []}'
+rules='[]'
+
+# sonarqube does not allow us to grab more than 10k issues, but most of them are information exposure which is not even
+# considered by benchmark so let's just get all relevant rules and receive results for only those rules
+
+rules_count=$(curl --silent -u "$sonar_token:" "$sonar_host/api/rules/search?p=1&ps=1" | jq -r '.total')
+page=1
+
+while (((page - 1) * 500 < rules_count)); do
+  rules=$(echo "$rules" | jq ". += $(curl --silent -u "$sonar_token:" "$sonar_host/api/rules/search?p=$page&ps=500" | jq '.rules | map( .key ) | map( select(. | contains("java:") ) )')")
+  page=$((page+1))
+done
+
+rules=$(echo "$rules" | jq '. | join(",")' | sed 's/java:S1989,//')
+
+issues_count=$(curl --silent -u "$sonar_token:" "$sonar_host/api/issues/search?p=1&ps=1&types=VULNERABILITY&componentKeys=$sonar_project&rules=$rules" | jq -r '.paging.total')
+page=1
+
+while (((page - 1) * 500 < issues_count)); do
+  issues_page=$(curl --silent -u "$sonar_token:" "$sonar_host/api/issues/search?types=VULNERABILITY&p=$page&ps=500&componentKeys=$sonar_project&rules=$rules" | jq '.issues')
+
+  result=$(echo "$result" | jq ".issues += $issues_page")
+  page=$((page+1))
+done
+
+hotspot_count=$(curl --silent -u "$sonar_token:" "$sonar_host/api/hotspots/search?projectKey=benchmark&p=1&ps=1" | jq -r '.paging.total')
+page=1
+
+while (((page - 1) * 500 < hotspot_count)); do
+  result=$(echo "$result" | jq ".hotspots += $(curl --silent -u "$sonar_token:" "$sonar_host/api/hotspots/search?projectKey=$sonar_project&p=$page&ps=500" | jq '.hotspots')")
+  page=$((page+1))
+done
+
+echo "$result" > "$result_file"
```

Comment:
When I run this, this happens: Dave.Wichers@USC02FJ2VXMD6T Benchmark % scripts/runSonarQube.sh 
[INFO] BuildTimeEventSpy is registered.
[INFO] Scanning for projects...
Downloading from central: https://repo.maven.apache.org/maven2/org/codehaus/mojo/sonar-maven-plugin/maven-metadata.xml
[WARNING] Could not transfer metadata org.codehaus.mojo:sonar-maven-plugin/maven-metadata.xml from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/codehaus/mojo/sonar-maven-plugin/maven-metadata.xml
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.859 s
[INFO] Finished at: 2021-09-14T14:38:43-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Error resolving version for plugin 'org.codehaus.mojo:sonar-maven-plugin' from the repositories [local (/Users/Dave.Wichers/.m2/repository), central (https://repo.maven.apache.org/maven2)]: Plugin not found in any plugin repository -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginVersionResolutionException
^C

How is this supposed to work?

Reply:
This is a copy-paste-command from SonarQube. Did you update the credentials in the file?

Reply:
If credentials need to be changed in the script, then the script should come with NO credentials, detect when they have not been provided, and provide instructions on what to do.  Look at the runBenchmark_wContrast.sh as an example of this. It's probably better if the credentials sit in a separate file (again like Contrast), so user's can edit the credential file separately. This allows them to pick up improvements to the script via git pull, without merge conflicts.

Reply:
done. please check :-)

Reply:
is there a SaaS instance of SonarQube we can use? If so, we should 'suggest' those in the SonarQubeCredential file. I'm not super interested in setting up a local SonarQube instance as that's a pain.  If a user can use a SaaS instance, and simply sign up for an account to use first, that would make it really easy for them to scan Benchmark w/SonarQube.

Reply:
I just used the quickstart command from here: https://docs.sonarqube.org/latest/setup/get-started-2-minutes/ with no configuration.

Reply:
I just used https://docs.sonarqube.org/latest/setup/get-started-2-minutes/ without any configuration. Just wanted to have a running script.

Reply:
That's great. Given that it is using docker, can you update your script to launch the Docker VM automatically, and then interact with it. Then you could embed the default credentials right into it. Also, you could put instructions on how to change the script to use your own SonarQube instance if you don't want to use Docker. Maybe create a separate SonarQubeUsageInstructions.txt file that explains its default behavior, and how to make these changes if you want to.